contracts/dao/StakingVault.sol

```diff
@@ -0,0 +1,206 @@
+pragma solidity ^0.8.10;
+
+import "./interfaces/IStaking.sol";
+import "./interfaces/IStakingVault.sol";
+import "./interfaces/IV2Wrapper.sol";
+import "./interfaces/IDistributor.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+contract StakingVault is OwnableUpgradeable, ReentrancyGuardUpgradeable, PausableUpgradeable {
+    using SafeERC20 for IERC20;
+
+    // staking address
+    address public staking;
+    // reward token
+    address public rewardToken;
+    // distributor => allocated amount
+    mapping(address => uint256) public allocated;
+    // fees
+    uint256 public fees;
+    // fee rate
+    uint256 public feeRate;
+    // max fee rate
+    uint256 public constant MAX_FEE = 10000;
+    // fee receiver
+    address public feeReceiver;
+    // lp proxy address
+    address public lpProxy;
+    // pauser address
+    address public pauser;
+
+    event AddRewards(address distributor, uint256 amount, uint256 fee);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+      * @dev initialize contract
+      * @param _owner owner address
+      * @param _rewardToken reward token address
+      * @param _feeReceiver fee receiver address
+      */
+    function initialize(
+        address _owner,
+        address _rewardToken,
+        address _feeReceiver
+    ) public initializer {
+        require(_owner != address(0), "owner cannot be zero address");
+        require(_rewardToken != address(0), "rewardToken cannot be zero address");
+        require(_feeReceiver != address(0), "feeReceiver cannot be zero address");
+        __Ownable_init();
+        transferOwnership(_owner);
+
+        rewardToken = _rewardToken;
+        feeReceiver = _feeReceiver;
+    }
+
+    modifier onlyPauser() {
+        require(msg.sender == pauser, "Only pauser can call this function");
+        _;
+    }
+
+    modifier onlyStaking() {
+        require(msg.sender == staking, "Only staking contract can call this function");
+        _;
+    }
+
+    modifier onlyLpProxy() {
+        require(msg.sender == lpProxy, "Only lp proxy can call this function");
+        _;
+    }
+
+    /**
+      * @dev send rewards to distributor
+      * @param distributor distributor address
+      * @param amount reward amount
+      */
+    function sendRewards(address distributor, uint256 amount) external onlyStaking {
+        IERC20(rewardToken).safeTransferFrom(staking, address(this), amount);
+
+        uint256 rewardAmount = amount;
+        uint256 fee;
+        if (feeRate > 0) {
+            fee = Math.mulDiv(amount, feeRate, MAX_FEE);
```

Comment:
Probably, it would be better to limit the feeRate to something more like 50-70% of Max. You can have two variables: FEE_PRECISION which is set to 10000 and then MAX_FEE which is set to let's say 7000. Your formula here will be: fee = Math.mulDiv(amount, feeRate, FEE_PRECISION); and on the setter you can check fee is <= MAX_FEE

Currently if somehow the fee is set to MAX_FEE then all the amount will be taken as fees, it would also help in building trust.